source/features/useful-not-found-page.tsx

```diff
@@ -107,19 +123,55 @@ function init(): false | void {
 	for (let i = bar.children.length - 2; i >= 0; i--) {
 		void checkAnchor(bar.children[i] as HTMLAnchorElement);
 	}
+}
 
-	if (['tree', 'blob'].includes(parts[2])) {
-		// Object might be 410 Gone
-		void addCommitHistoryLink(bar);
+async function showHelpfulLinks(): Promise<void> {
+	// What we should try to do to be helpful:
+	//  1. Check if the file is on the default branch
+	//  2. Check the current branch's history for the file (was it deleted or renamed)
+
+	const urlToFileOnDefaultBranch = await getUrlToFileOnDefaultBranch();
+	if (urlToFileOnDefaultBranch) {
+		select('h2')!.after(
+			<p className="container mt-4 text-center">
+				<a href={urlToFileOnDefaultBranch}>This {getType()}</a> exists on the default branch.
+			</p>,
+		);
+		return;
 	}
 
-	if (['tree', 'blob', 'edit'].includes(parts[2])) {
-		// File might not be available on the current branch
-		// GitHub already redirects /tree/ and /blob/ natively
-		void addDefaultBranchLink(bar);
+	const change = await getLatestChangeToFile();
+	if (change) {
```

Comment:
Prefer early returns to reduce indentation
```suggestion
	if (!change) {
		return;
	}

```